### ISSUE TITLE
SUPPA2 clustering analysis on diffSPlicing with nan value

### ISSUE THREAD

**Issue by Mvirale** [comprna/SUPPA#220]
Hi, 

I'm trying to run cluster analysisi on diffsplicing output and i'm encountering somne issue related to 'nan' output. 

After running the analsysis on the dpsi and Psivec files i'm getting this error : 'ERROR:lib.cluster_tools:Unknown error: (<class 'ValueError'>, ValueError('Input X contains NaN.\nDBSCAN does not accept missing values encoded as NaN natively. For supervised learning, you might want to consider sklearn.ensemble.HistGradientBoostingClassifier and Regressor which accept missing values encoded as NaNs natively. Alternatively, it is possible to preprocess the data, for instance by using an imputer transformer in a pipeline or drop samples with missing values. See [https://scikit-learn.org/stable/modules/impute.html](vscode-file://vscode-app/c:/Users/vital/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-sandbox/workbench/workbench.html) You can find a list of all estimators that handle NaN values at the following page: [https://scikit-learn.org/stable/modules/impute.html#estimators-that-handle-nan-values](vscode-file://vscode-app/c:/Users/vital/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-sandbox/workbench/workbench.html)'), <traceback object at 0x14dd9a479480>)
' what can i do ? preprocess the file ? what i could conerrt the nan to ?' 

To overcome this issue i tried to force removal of 'nan' using --nan 0 and --lower-bound 0' but nothing changed. I then tried to modify the cluster_tool script to deal with 'nan' value, usign the SimpleImputer function of sklearn : 

@@ top of file, with the other imports
-from sklearn.cluster import DBSCAN
+from sklearn.cluster import DBSCAN
+from sklearn.impute import SimpleImputer
@@
 def DBSCAN_cluster(psi_matrix, eventid_lst, dist, minpts, metric):
 
     # Setting logging preferences
     logger = logging.getLogger(__name__)
+
+    # ---- NEW: impute missing PSI values by column mean ----
+    import numpy as _np
+    X = _np.array(psi_matrix, dtype=float)
+    imputer = SimpleImputer(missing_values=_np.nan, strategy='mean')
+    X = imputer.fit_transform(X)
+    # --------------------------------------------------------
 
     # The metric "cosine" only works with algorithm="brute"
     if metric == "cosine":
         alg = 'brute'
@@
-    try:
-        db = DBSCAN(eps=dist, min_samples=minpts, metric=metric, algorithm=alg).fit(psi_matrix)
+    try:
+        db = DBSCAN(eps=dist, min_samples=minpts, metric=metric, algorithm=alg).fit(X)
         labels = db.labels_
     except:
         logger.error("Unknown error: {}".format(sys.exc_info()))'. this is the script that i used. 

Now the script works and can take as input 'nan' values.  However, i was wondering how thsi can affect the dowstream analysisi of the clustering tool. Can you please give me some feedbacks? Also what is the better option? using the mean or convertig the 'nan' value to 0 ? 

Thanks in advance

Best

Matteo
